Add hop annotation support

SaiedKazemi opened the pull request:
This commit adds hop annotation support (hopannotation1 datatype)
to traceroute-caller.

For details, run the following command:

  go doc -all hopannotation

The contents of a typical hop annotation file looks like the
following:
```
  {
    "ID": "20210824_2d638011386a_91.189.91.39",
    "Timestamp": "2021-08-24T15:27:59Z",
    "Annotations": {
      "Geo": {
	"ContinentCode": "NA",
	"CountryCode": "US",
	"CountryName": "United States",
	"Subdivision1ISOCode": "MA",
	"Subdivision1Name": "Massachusetts",
	"MetroCode": 506,
	"City": "Boston",
	"PostalCode": "02112",
	"Latitude": 42.3562,
	"Longitude": -71.0631,
	"AccuracyRadiusKm": 200
      },
      "Network": {
	"CIDR": "91.189.88.0/21",
	"ASNumber": 41231,
	"ASName": "Canonical Group Limited",
	"Systems": [
	  {
	    "ASNs": [
	      41231
	    ]
	  }
	]
      }
    }
  }
```
The changes were successfully tested with:

  go test ./...
  go test -race ./...
  docker-compose up

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/m-lab/traceroute-caller/120)
<!-- Reviewable:end -->

Commit:
Add hop annotation support

This commit adds hop annotation support (hopannotation1 datatype)
to traceroute-caller.

For details, run the following command:

  go doc -all hopannotation

The contents of a typical hop annotation file looks like the
following:

  {
    "ID": "20210824_2d638011386a_91.189.91.39",
    "Timestamp": "2021-08-24T15:27:59Z",
    "Annotations": {
      "Geo": {
	"ContinentCode": "NA",
	"CountryCode": "US",
	"CountryName": "United States",
	"Subdivision1ISOCode": "MA",
	"Subdivision1Name": "Massachusetts",
	"MetroCode": 506,
	"City": "Boston",
	"PostalCode": "02112",
	"Latitude": 42.3562,
	"Longitude": -71.0631,
	"AccuracyRadiusKm": 200
      },
      "Network": {
	"CIDR": "91.189.88.0/21",
	"ASNumber": 41231,
	"ASName": "Canonical Group Limited",
	"Systems": [
	  {
	    "ASNs": [
	      41231
	    ]
	  }
	]
      }
    }
  }

The changes were successfully tested with:

  go test ./...
  go test -race ./...
  docker-compose up

gfr10598 reviewed (changes requested):
> Reviewed 1 of 5 files at r1.
> *[Reviewable](https://reviewable.io/reviews/m-lab/traceroute-caller/120)* status: 1 change requests, 0 of 1 approvals obtained (waiting on @SaiedKazemi)

___
*[.gitignore, line 1 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-MhyrhVt2Ms0mtOEUUbS:-MhyrhVt2Ms0mtOEUUbT:b-r0vjj2) ([raw file](https://github.com/m-lab/traceroute-caller/blob/ddba1bdfb3720a796d2f243a001ac440ba6fdf5a/.gitignore#L1)):*
> ```ignore
> local
> ```

Not clear to me that .gitignore should be in source code control.  Prefer that you remove it.
___
*[connectionlistener/connectionlistener.go, line 34 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-Mhz-o1H1tZJP6yKV9h6:-Mhz-o1H1tZJP6yKV9h7:b-q96t4w) ([raw file](https://github.com/m-lab/traceroute-caller/blob/ddba1bdfb3720a796d2f243a001ac440ba6fdf5a/connectionlistener/connectionlistener.go#L34)):*
> ```Go
> 	ipCache      *ipcache.RecentIPCache
> 	creator      connection.Creator
> 	hopAnnotator *hopannotation.HopCache
> ```

I think the hopCaches for yesterday and today should be managed here.  There should be a single hopcache reference, which should be replaced with a new empty HopCache at the time of the first event after midnight.  The HopCache should include a date field (or method) that indicates what date it covers.
___
*[connectionlistener/connectionlistener.go, line 84 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-Mhz-_N6EW_IoEPn-Zs3:-Mhz-_N6EW_IoEPn-Zs4:b1vvwrk) ([raw file](https://github.com/m-lab/traceroute-caller/blob/ddba1bdfb3720a796d2f243a001ac440ba6fdf5a/connectionlistener/connectionlistener.go#L84)):*
> ```Go
> }
> 
> func (cl *connectionListener) traceAnnotateAndArchive(ctx context.Context, conn connection.Connection) {
> ```

At this point, grab a reference to the HopCache.  If time.Now is not the same date as the HopCache, then replace cl.HopCache with a new one.

  hopCache := cl.GetTodaysHopCache()  // thread safe - replaces HopCache with new one if the date has just changed.
___
*[connectionlistener/connectionlistener.go, line 100 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-Mhz-MFG86uz-Iis6Ty6:-Mhz-MFH0-1aapLFPlkC:b-imxpd5) ([raw file](https://github.com/m-lab/traceroute-caller/blob/ddba1bdfb3720a796d2f243a001ac440ba6fdf5a/connectionlistener/connectionlistener.go#L100)):*
> ```Go
> 		return
> 	}
> ```

At this point, we can determine whether to use yesterday's HopCache, or today's.  We can now check if the value of priorHopCache is correct, or not.  

  if output.CycleStart.StartTime different date than priorHopCache {
    hopCache = cl.GetTodaysHopCache()  // Which should be the new one.
  }
___
*[connectionlistener/connectionlistener.go, line 105 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-MhyuyXw-M0WhdtIwzao:-MhyuyXw-M0WhdtIwzap:b82o0v8) ([raw file](https://github.com/m-lab/traceroute-caller/blob/ddba1bdfb3720a796d2f243a001ac440ba6fdf5a/connectionlistener/connectionlistener.go#L105)):*
> ```Go
> 		log.Printf("failed to annotate some or all hops (errors: %+v)\n", allErrs)
> 	}
> 	if annotations != nil {
> ```

Also check for non-zero length?
___
*[hopannotation/hopannotation.go, line 74 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-Mhyxr4X8AEhA-qMxLU2:-Mhyxr4X8AEhA-qMxLU3:bur04ek) ([raw file](https://github.com/m-lab/traceroute-caller/blob/ddba1bdfb3720a796d2f243a001ac440ba6fdf5a/hopannotation/hopannotation.go#L74)):*
> ```Go
> 
> // HopCache is the cache of hop annotations.
> type HopCache struct {
> ```

This structure makes it hard to reason about concurrency, and makes the code complicated.

Instead, HopCache should handle only one map, and there should be more than one HopCache.  Ideally, yesterday's HopCache should only have references from traces in flight, and once all the traces complete and the hops are annotated, it could be garbage collected.

However, since we don't know until we get the traceroute json back what the cyclestart time will be, it is hard to hold a reference to the correct cache.  The traceroute might be launched at 23:59.99999, and cyclestart might be 00:00.000001. 

Thinking out loud:
  If we provide Trace with an indirect reference to the currentHopCache (with a Get method) and a direct reference to the recentHopCache that is determined prior to calling Trace, then trace can determine which one to use based on cyclestart time.  Most of the time, the currentHopCache.Get() will return the same value as recentHopCache, but near midnight they might be different.

  Then Trace uses the cyclestart time to determine which cache should be used for the hop annotations.
___
*[hopannotation/hopannotation.go, line 76 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-Mhys4Bl7Rd2H644pKSY:-Mhys4BmEsVcG5R_LuvB:b-6txiph) ([raw file](https://github.com/m-lab/traceroute-caller/blob/ddba1bdfb3720a796d2f243a001ac440ba6fdf5a/hopannotation/hopannotation.go#L76)):*
> ```Go
> type HopCache struct {
> 	hops       map[string]entryState // hop addresses being handled or already handled
> 	oldHops    map[string]entryState // old (yesterday's) hops
> ```

Does this need to be maintained, or can the old hop map just be orphaned?  If we have to keep it around, please document why.
___
*[hopannotation/hopannotation.go, line 80 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-MhytNKc5ZI3amg-pg-6:-MhytNKc5ZI3amg-pg-7:bnsy8yc) ([raw file](https://github.com/m-lab/traceroute-caller/blob/ddba1bdfb3720a796d2f243a001ac440ba6fdf5a/hopannotation/hopannotation.go#L80)):*
> ```Go
> 	annotator  ipservice.Client      // function for getting hop annotations
> 	outputPath string                // path to directory for writing hop annotations
> 	hour       int32                 // the hour when cache clearer last checked time
> ```

The time checking can be done by every call to the cache.
___
*[hopannotation/hopannotation.go, line 116 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-Mhytw-wDoMibYFmAhGa:-Mhytw-wDoMibYFmAhGb:b80dwac) ([raw file](https://github.com/m-lab/traceroute-caller/blob/ddba1bdfb3720a796d2f243a001ac440ba6fdf5a/hopannotation/hopannotation.go#L116)):*
> ```Go
> 			}
> 			hour := now.Hour()
> 			if hour < int(atomic.LoadInt32(&hc.hour)) {
> ```

Should probably use a swap rather than a load, to completely avoid races.
___
*[hopannotation/hopannotation.go, line 249 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-MhysJ9M3KGjwm0v2cii:-MhysJ9M3KGjwm0v2cij:b-z57dm7) ([raw file](https://github.com/m-lab/traceroute-caller/blob/ddba1bdfb3720a796d2f243a001ac440ba6fdf5a/hopannotation/hopannotation.go#L249)):*
> ```Go
> // setState sets the state of the given hop entry in the cache to the
> // given state while holding the hop cache lock.
> func (hc *HopCache) setState(hop string, hopState entryState) {
> ```

It would be much better if the state handling is done by the handler using a reference on the entry.  It shouldn't be necessary to access this through the cache.
___
*[hopannotation/hopannotation.go, line 269 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-MhyxROR8z5wRiW6fyLE:-MhyxROR8z5wRiW6fyLF:bnw7g4q) ([raw file](https://github.com/m-lab/traceroute-caller/blob/ddba1bdfb3720a796d2f243a001ac440ba6fdf5a/hopannotation/hopannotation.go#L269)):*
> ```Go
> 	if !ok {
> 		// Check the old cache.
> 		state, ok = hc.oldHops[hop]
> ```

This is unnecessarily complicated.  old vs new map should be determined based on cyclestart time, not by searching both caches.

It would be best if the annotateHop function worked on a single cache, either yesterdays or todays.  I'll add comments elsewhere to clarify this.


<!-- Sent from Reviewable.io -->

gfr10598 commented:
___
*[hopannotation/hopannotation.go, line 168 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-Mhz2zfk451hKe5fyYtG:-Mhz2zfk451hKe5fyYtH:bjtqaxg) ([raw file](https://github.com/m-lab/traceroute-caller/blob/ddba1bdfb3720a796d2f243a001ac440ba6fdf5a/hopannotation/hopannotation.go#L168)):*
> ```Go
> 		hc.hopsLock.Lock()
> 		if _, ok := hc.hops[hop]; !ok {
> 			hc.hops[hop] = inserted
> ```

Please add a metric.
I'm inclined to think we need to be better about adding metrics before pushing to prod.  Since this is a candidate for prod, it should have prom metrics.


<!-- Sent from Reviewable.io -->

gfr10598 reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/m-lab/traceroute-caller/120)* status: 1 change requests, 0 of 1 approvals obtained (waiting on @SaiedKazemi)

___
*[hopannotation/hopannotation.go, line 46 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-Mhz2nea9EPSKsjPqAfc:-Mhz2nea9EPSKsjPqAfd:biptuh2) ([raw file](https://github.com/m-lab/traceroute-caller/blob/ddba1bdfb3720a796d2f243a001ac440ba6fdf5a/hopannotation/hopannotation.go#L46)):*
> ```Go
> 
> const (
> 	inserted entryState = iota
> ```

I like these states, but it would be more useful to have metrics than to keep these here.  Both is fine, of course.


<!-- Sent from Reviewable.io -->

Commit:
Remove debug code from hopannotation

SaiedKazemi commented:
___
*[.gitignore, line 1 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-MhyrhVt2Ms0mtOEUUbS:-Mhzr3BR5tNp_EaQMUmG:b-ofb3px) ([raw file](https://github.com/m-lab/traceroute-caller/blob/ddba1bdfb3720a796d2f243a001ac440ba6fdf5a/.gitignore#L1)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Not clear to me that .gitignore should be in source code control.  Prefer that you remove it.
</blockquote></details>

.gitignore is tracked to share the rule list with other users.  There are more than 100 M-Lab repos that track .gitignore.


<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
___
*[connectionlistener/connectionlistener.go, line 34 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-Mhz-o1H1tZJP6yKV9h6:-Mhzs1Ze6-YUaptDMMk0:b359dd2) ([raw file](https://github.com/m-lab/traceroute-caller/blob/ddba1bdfb3720a796d2f243a001ac440ba6fdf5a/connectionlistener/connectionlistener.go#L34)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

I think the hopCaches for yesterday and today should be managed here.  There should be a single hopcache reference, which should be replaced with a new empty HopCache at the time of the first event after midnight.  The HopCache should include a date field (or method) that indicates what date it covers.
</blockquote></details>

After reading through your comments I realized that the "sanity check" code that I had added during debugging has caused confusion.

There is technically no need to keep yesterday's cache.  It can be orphaned by Clear() at midnight and garbage collected.

I have now removed oldHops and setState().


<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
___
*[connectionlistener/connectionlistener.go, line 84 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-Mhz-_N6EW_IoEPn-Zs3:-MhztNB_AJv3KiQtUlu6:bdwfgce) ([raw file](https://github.com/m-lab/traceroute-caller/blob/ddba1bdfb3720a796d2f243a001ac440ba6fdf5a/connectionlistener/connectionlistener.go#L84)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

At this point, grab a reference to the HopCache.  If time.Now is not the same date as the HopCache, then replace cl.HopCache with a new one.

  hopCache := cl.GetTodaysHopCache()  // thread safe - replaces HopCache with new one if the date has just changed.
</blockquote></details>

Per my earlier comment, there's no need to do this.


<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
___
*[connectionlistener/connectionlistener.go, line 100 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-Mhz-MFG86uz-Iis6Ty6:-MhztTpKAdeyA3R2Yvw_:bue3ebm) ([raw file](https://github.com/m-lab/traceroute-caller/blob/ddba1bdfb3720a796d2f243a001ac440ba6fdf5a/connectionlistener/connectionlistener.go#L100)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

At this point, we can determine whether to use yesterday's HopCache, or today's.  We can now check if the value of priorHopCache is correct, or not.  

  if output.CycleStart.StartTime different date than priorHopCache {
    hopCache = cl.GetTodaysHopCache()  // Which should be the new one.
  }


</blockquote></details>

Per my earlier comment, there is no need to do this.


<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
___
*[connectionlistener/connectionlistener.go, line 105 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-MhyuyXw-M0WhdtIwzao:-MhztXqr3Lp-_Z7_Mt1A:b-csemez) ([raw file](https://github.com/m-lab/traceroute-caller/blob/ddba1bdfb3720a796d2f243a001ac440ba6fdf5a/connectionlistener/connectionlistener.go#L105)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Also check for non-zero length?
</blockquote></details>

If err is nil, annotations is not zero length.  Even if it is zero length, it doesn't cause any issues.  But I added the check for non-zero length as you have suggested.


<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
___
*[hopannotation/hopannotation.go, line 46 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-Mhz2nea9EPSKsjPqAfc:-Mhzu0SM3_6URWRcCAi-:b-husu1t) ([raw file](https://github.com/m-lab/traceroute-caller/blob/ddba1bdfb3720a796d2f243a001ac440ba6fdf5a/hopannotation/hopannotation.go#L46)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

I like these states, but it would be more useful to have metrics than to keep these here.  Both is fine, of course.
</blockquote></details>

These for sanity checks.  I have removed them and added metrics instead.


<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
___
*[hopannotation/hopannotation.go, line 74 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-Mhyxr4X8AEhA-qMxLU2:-MhzuDwZ0IyXPSh8yUFc:b4z7m2o) ([raw file](https://github.com/m-lab/traceroute-caller/blob/ddba1bdfb3720a796d2f243a001ac440ba6fdf5a/hopannotation/hopannotation.go#L74)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

This structure makes it hard to reason about concurrency, and makes the code complicated.

Instead, HopCache should handle only one map, and there should be more than one HopCache.  Ideally, yesterday's HopCache should only have references from traces in flight, and once all the traces complete and the hops are annotated, it could be garbage collected.

However, since we don't know until we get the traceroute json back what the cyclestart time will be, it is hard to hold a reference to the correct cache.  The traceroute might be launched at 23:59.99999, and cyclestart might be 00:00.000001. 

Thinking out loud:
  If we provide Trace with an indirect reference to the currentHopCache (with a Get method) and a direct reference to the recentHopCache that is determined prior to calling Trace, then trace can determine which one to use based on cyclestart time.  Most of the time, the currentHopCache.Get() will return the same value as recentHopCache, but near midnight they might be different.

  Then Trace uses the cyclestart time to determine which cache should be used for the hop annotations.
</blockquote></details>

Per my earlier comments, I have removed oldHops and entryState.  The cache is now map[string]bool.


<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
___
*[hopannotation/hopannotation.go, line 76 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-Mhys4Bl7Rd2H644pKSY:-MhzuTFY7NrFXgubmUEC:b-wj1usa) ([raw file](https://github.com/m-lab/traceroute-caller/blob/ddba1bdfb3720a796d2f243a001ac440ba6fdf5a/hopannotation/hopannotation.go#L76)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Does this need to be maintained, or can the old hop map just be orphaned?  If we have to keep it around, please document why.
</blockquote></details>

There's no oldHops anymore.


<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
___
*[hopannotation/hopannotation.go, line 80 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-MhytNKc5ZI3amg-pg-6:-Mhzub6i4IHIvsa55k1t:b-yxg7cn) ([raw file](https://github.com/m-lab/traceroute-caller/blob/ddba1bdfb3720a796d2f243a001ac440ba6fdf5a/hopannotation/hopannotation.go#L80)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

The time checking can be done by every call to the cache.
</blockquote></details>

I have reworked the code after removing sanity checks.


<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
___
*[hopannotation/hopannotation.go, line 116 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-Mhytw-wDoMibYFmAhGa:-Mhzuusr9391djWQ1ZXg:bsqchpe) ([raw file](https://github.com/m-lab/traceroute-caller/blob/ddba1bdfb3720a796d2f243a001ac440ba6fdf5a/hopannotation/hopannotation.go#L116)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Should probably use a swap rather than a load, to completely avoid races.
</blockquote></details>

As the comment says, this is solely needed for better code coverage via package testing and can be removed if it's an issue.


<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
___
*[hopannotation/hopannotation.go, line 168 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-Mhz2zfk451hKe5fyYtG:-Mhzv8Eh3aaV2N78lmAs:b-896fix) ([raw file](https://github.com/m-lab/traceroute-caller/blob/ddba1bdfb3720a796d2f243a001ac440ba6fdf5a/hopannotation/hopannotation.go#L168)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Please add a metric.
I'm inclined to think we need to be better about adding metrics before pushing to prod.  Since this is a candidate for prod, it should have prom metrics.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
___
*[hopannotation/hopannotation.go, line 249 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-MhysJ9M3KGjwm0v2cii:-MhzvAr-D-FqUJFcJSQL:bqh3d9y) ([raw file](https://github.com/m-lab/traceroute-caller/blob/ddba1bdfb3720a796d2f243a001ac440ba6fdf5a/hopannotation/hopannotation.go#L249)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

It would be much better if the state handling is done by the handler using a reference on the entry.  It shouldn't be necessary to access this through the cache.
</blockquote></details>

All of this is gone.


<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
___
*[hopannotation/hopannotation.go, line 269 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-MhyxROR8z5wRiW6fyLE:-MhzvEaj5C1yn-ybGBoS:bb8end5) ([raw file](https://github.com/m-lab/traceroute-caller/blob/ddba1bdfb3720a796d2f243a001ac440ba6fdf5a/hopannotation/hopannotation.go#L269)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

This is unnecessarily complicated.  old vs new map should be determined based on cyclestart time, not by searching both caches.

It would be best if the annotateHop function worked on a single cache, either yesterdays or todays.  I'll add comments elsewhere to clarify this.
</blockquote></details>

All of it is gone.


<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
PTAL.





<!-- Sent from Reviewable.io -->

gfr10598 reviewed (changes requested):
Made a bunch of comments, but let me summarize here:
a.  Much better!  Thanks!
b.  I think there is a race with the map clearing, but I'm ok with adding a todo/issue, and proceeding with the existing design, unless you want to fix it now.  I'd likely take on the issue sometime in the next few weeks.
c.  There are a few other nits and suggestions, and some code that I haven't completely reviewed yet.

Please let me know your preference.  If there are things you can respond to early, then I'll take another pass this afternoon.

> *[Reviewable](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-:-Mi23p1s4Hz5BTUba_J9:b-k5y3x3)* status: 1 change requests, 0 of 1 approvals obtained (waiting on @gfr10598 and @SaiedKazemi)

___
*[hopannotation/hopannotation.go, line 168 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-Mhz2zfk451hKe5fyYtG:-Mi1W3BjC8AH_1Zw4wbM:b-hay2ia) ([raw file](https://github.com/m-lab/traceroute-caller/blob/ddba1bdfb3720a796d2f243a001ac440ba6fdf5a/hopannotation/hopannotation.go#L168)):*
<details><summary><i>Previously, SaiedKazemi (Saied Kazemi) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Thanks!!
___
*[hopannotation/hopannotation.go, line 51 at r2](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-Mi1Wz7k0eJx7Wt2hBPD:-Mi1Wz7k0eJx7Wt2hBPE:b-oa81z9) ([raw file](https://github.com/m-lab/traceroute-caller/blob/b15eddf427b3a2cb9379e05fb143c34fb2e5f524/hopannotation/hopannotation.go#L51)):*
> ```Go
> 			Help: "The number of hop cache operations",
> 		},
> 		[]string{"type", "operation"},
> ```

BTW, I'm a fan of good labels, but IIUC, they add to the metrics footprint multiplicatively, so it is good to be cautious about the dimensionality.  In other words, will we get good value from both labels here?  I've added labels in etl pipeline metrics that I later ended up regretting.
___
*[hopannotation/hopannotation.go, line 64 at r2](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-Mi1dfnY-ZBPsCJ11vKV:-Mi1dfnY-ZBPsCJ11vKW:b13t28f) ([raw file](https://github.com/m-lab/traceroute-caller/blob/b15eddf427b3a2cb9379e05fb143c34fb2e5f524/hopannotation/hopannotation.go#L64)):*
> ```Go
> 
> 	// Package testing aid.
> 	tickerDuration = int32(60 * 1000) // ticker duration in milliseconds for cache clearer
> ```

You treat this elsewhere as a time.Duration, which is in nanoseconds.
___
*[hopannotation/hopannotation.go, line 108 at r2](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-Mi1d3sX87RD5ISNhObW:-Mi1d3sYAssmhVei_wrn:b-frejl8) ([raw file](https://github.com/m-lab/traceroute-caller/blob/b15eddf427b3a2cb9379e05fb143c34fb2e5f524/hopannotation/hopannotation.go#L108)):*
> ```Go
> 	// a race condition with package testing code.
> 	go func() {
> 		d := time.Duration(atomic.LoadInt32(&tickerDuration))
> ```


Probably better to pass tickerDuration into the goroutine as a parameter.  I think that avoids the need for atomic access.

(Probably moot due to design flaw).
___
*[hopannotation/hopannotation.go, line 109 at r2](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-Mi1dwN76kJlPt_nwXWB:-Mi1dwN76kJlPt_nwXWC:b1g9e91) ([raw file](https://github.com/m-lab/traceroute-caller/blob/b15eddf427b3a2cb9379e05fb143c34fb2e5f524/hopannotation/hopannotation.go#L109)):*
> ```Go
> 	go func() {
> 		d := time.Duration(atomic.LoadInt32(&tickerDuration))
> 		ticker := time.NewTicker(d * time.Millisecond)
> ```

Really confusing to treat a Duration as something other than a nanosecond time.Duration value.  Please do the scaling where you create the Duration above, and rename tickerIntervalMilliseconds.   (Probably moot due to design flaw).
___
*[hopannotation/hopannotation.go, line 116 at r2](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-Mi1cS0F9nkeLjejDQ-W:-Mi1cS0F9nkeLjejDQ-X:b-8zti4b) ([raw file](https://github.com/m-lab/traceroute-caller/blob/b15eddf427b3a2cb9379e05fb143c34fb2e5f524/hopannotation/hopannotation.go#L116)):*
> ```Go
> 			}
> 			hour := now.Hour()
> 			if hour < int(atomic.LoadInt32(&hc.hour)) {
> ```

Is this backwards?  Seems like it should clear the cache if time.Now() > 
BTW, I've had some success testing this kind of code using monkey patch to patch the time.Now function.
___
*[hopannotation/hopannotation.go, line 128 at r2](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-Mi1cfVF9-4ez3JQ5Uzm:-Mi1cfVF9-4ez3JQ5Uzn:bng12gm) ([raw file](https://github.com/m-lab/traceroute-caller/blob/b15eddf427b3a2cb9379e05fb143c34fb2e5f524/hopannotation/hopannotation.go#L128)):*
> ```Go
> // than the current cache.  The current cache is retained as old cache
> // to allow for active annotations to finish.
> func (hc *HopCache) Clear() {
> ```

Maybe call this Reset, since it creates a new map, rather than clearing the old map?  Don't feel strongly about it.
___
*[hopannotation/hopannotation.go, line 146 at r2](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-Mi1VwJV1-yjFxZhgQWd:-Mi1VwJV1-yjFxZhgQWe:b2nm4hk) ([raw file](https://github.com/m-lab/traceroute-caller/blob/b15eddf427b3a2cb9379e05fb143c34fb2e5f524/hopannotation/hopannotation.go#L146)):*
> ```Go
> 	for _, hop := range hops {
> 		if net.ParseIP(hop).String() == "<nil>" {
> 			allErrs = append(allErrs, fmt.Errorf("%w: %v", errParseHopIP, hop))
> ```

Since this list of errors will be returned to caller, and method is exported, the error types should also be exported.
___
*[hopannotation/hopannotation.go, line 156 at r2](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-Mi1efuR4XxwzP6q9bzX:-Mi1efuR4XxwzP6q9bzY:baqa39q) ([raw file](https://github.com/m-lab/traceroute-caller/blob/b15eddf427b3a2cb9379e05fb143c34fb2e5f524/hopannotation/hopannotation.go#L156)):*
> ```Go
> 	var newHops []string
> 	for _, hop := range hops {
> 		hc.hopsLock.Lock()
> ```

Probably better to hold the lock outside the loop.  
___
*[hopannotation/hopannotation.go, line 156 at r2](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-Mi1fL-0B47qGfZty_zq:-Mi1fL-0B47qGfZty_zr:b-htughz) ([raw file](https://github.com/m-lab/traceroute-caller/blob/b15eddf427b3a2cb9379e05fb143c34fb2e5f524/hopannotation/hopannotation.go#L156)):*
> ```Go
> 	var newHops []string
> 	for _, hop := range hops {
> 		hc.hopsLock.Lock()
> ```

I think there is a fundamental race between this loop, and the goroutine that replaces the map.  It will occasionally result in some of the hops being registered in yesterday's map, and others being registered in today's map.  If this results in duplicate entries, it is no big deal, but I think it can also result in a hop from yesterday's trace being recorded in today's map, so there won't be a hop recorded in the correct date partition, and the JOIN will later fail in the ETL pipeline.

So, I think you want to replace the map on demand, in this Annotate method, when the trace cyclestart time does not correspond to the date for the map.  This will then result in the same problem I mentioned in previous comment


<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
I can respond early.





<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
___
*[hopannotation/hopannotation.go, line 51 at r2](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-Mi1Wz7k0eJx7Wt2hBPD:-Mi2Gkl80g59TwK5Z10P:b-6hgv43) ([raw file](https://github.com/m-lab/traceroute-caller/blob/b15eddf427b3a2cb9379e05fb143c34fb2e5f524/hopannotation/hopannotation.go#L51)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

BTW, I'm a fan of good labels, but IIUC, they add to the metrics footprint multiplicatively, so it is good to be cautious about the dimensionality.  In other words, will we get good value from both labels here?  I've added labels in etl pipeline metrics that I later ended up regretting.
</blockquote></details>

This is my first experience with Prometheus metrics, so let's discuss interactively when we meet.


<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
___
*[hopannotation/hopannotation.go, line 64 at r2](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-Mi1dfnY-ZBPsCJ11vKV:-Mi2H1xgCWLh20HB4qPj:bcmpil9) ([raw file](https://github.com/m-lab/traceroute-caller/blob/b15eddf427b3a2cb9379e05fb143c34fb2e5f524/hopannotation/hopannotation.go#L64)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

You treat this elsewhere as a time.Duration, which is in nanoseconds.
</blockquote></details>

Made it more clear.


<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
___
*[hopannotation/hopannotation.go, line 108 at r2](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-Mi1d3sX87RD5ISNhObW:-Mi2IDo51bB-13_G0nQY:bn3yjtc) ([raw file](https://github.com/m-lab/traceroute-caller/blob/b15eddf427b3a2cb9379e05fb143c34fb2e5f524/hopannotation/hopannotation.go#L108)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>


Probably better to pass tickerDuration into the goroutine as a parameter.  I think that avoids the need for atomic access.

(Probably moot due to design flaw).
</blockquote></details>

It's not moot because package testing code needs to work with much shorter intervals in order to avoid long waits for midnight.  I am now passing it in.


<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
___
*[hopannotation/hopannotation.go, line 109 at r2](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-Mi1dwN76kJlPt_nwXWB:-Mi2Il0oAdWEw4gItoq1:buimrh6) ([raw file](https://github.com/m-lab/traceroute-caller/blob/b15eddf427b3a2cb9379e05fb143c34fb2e5f524/hopannotation/hopannotation.go#L109)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Really confusing to treat a Duration as something other than a nanosecond time.Duration value.  Please do the scaling where you create the Duration above, and rename tickerIntervalMilliseconds.   (Probably moot due to design flaw).
</blockquote></details>

It's not moot.  Per my earlier comment, I made it clear.


<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
___
*[hopannotation/hopannotation.go, line 116 at r2](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-Mi1cS0F9nkeLjejDQ-W:-Mi2Iwsf2sp69a0TPRiS:b28v0n1) ([raw file](https://github.com/m-lab/traceroute-caller/blob/b15eddf427b3a2cb9379e05fb143c34fb2e5f524/hopannotation/hopannotation.go#L116)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Is this backwards?  Seems like it should clear the cache if time.Now() > 
BTW, I've had some success testing this kind of code using monkey patch to patch the time.Now function.

</blockquote></details>

No, it's not backwards.  Hours go from 0 to 23 and then back to 0 again. Therefore, when the current hour is less than the previous hour, we have passed midnight.


<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
___
*[hopannotation/hopannotation.go, line 146 at r2](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-Mi1VwJV1-yjFxZhgQWd:-Mi2LMrz9_uKAzie3Pax:b-896fix) ([raw file](https://github.com/m-lab/traceroute-caller/blob/b15eddf427b3a2cb9379e05fb143c34fb2e5f524/hopannotation/hopannotation.go#L146)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Since this list of errors will be returned to caller, and method is exported, the error types should also be exported.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
___
*[hopannotation/hopannotation.go, line 156 at r2](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-Mi1fL-0B47qGfZty_zq:-Mi2Lt8WBPb6CisHGi8T:b-59w97g) ([raw file](https://github.com/m-lab/traceroute-caller/blob/b15eddf427b3a2cb9379e05fb143c34fb2e5f524/hopannotation/hopannotation.go#L156)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

I think there is a fundamental race between this loop, and the goroutine that replaces the map.  It will occasionally result in some of the hops being registered in yesterday's map, and others being registered in today's map.  If this results in duplicate entries, it is no big deal, but I think it can also result in a hop from yesterday's trace being recorded in today's map, so there won't be a hop recorded in the correct date partition, and the JOIN will later fail in the ETL pipeline.

So, I think you want to replace the map on demand, in this Annotate method, when the trace cyclestart time does not correspond to the date for the map.  This will then result in the same problem I mentioned in previous comment
</blockquote></details>

If the cache is reset while we are iterating this loop, it means that midnight has passed.  Therefore, the remaining hops in the hops slice will be annotated which is exactly the behavior we want.

I will add a comment explaining this in the code.


<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
___
*[hopannotation/hopannotation.go, line 156 at r2](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-Mi1efuR4XxwzP6q9bzX:-Mi2NFMEEhBzIKNqpjiP:bfn94d8) ([raw file](https://github.com/m-lab/traceroute-caller/blob/b15eddf427b3a2cb9379e05fb143c34fb2e5f524/hopannotation/hopannotation.go#L156)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Probably better to hold the lock outside the loop.  
</blockquote></details>

We don't have to lock the map for the whole duration.  Please see my comment below.


<!-- Sent from Reviewable.io -->

Commit:
Make changes suggested in code review

Commit:
Make the hop cache aware of date

gfr10598 reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/m-lab/traceroute-caller/120)* status: 1 change requests, 0 of 1 approvals obtained (waiting on @gfr10598 and @SaiedKazemi)

___
*[hopannotation/hopannotation.go, line 116 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-Mhytw-wDoMibYFmAhGa:-Mi2t9Bp8KSE2MrQUNd3:bdbbfl4) ([raw file](https://github.com/m-lab/traceroute-caller/blob/ddba1bdfb3720a796d2f243a001ac440ba6fdf5a/hopannotation/hopannotation.go#L116)):*
<details><summary><i>Previously, SaiedKazemi (Saied Kazemi) wrote…</i></summary><blockquote>

As the comment says, this is solely needed for better code coverage via package testing and can be removed if it's an issue.
</blockquote></details>

Ok - had to think about what you meant for a bit.  Now I'm curious why there is a race in testing.  It seems like the only goroutine that should access this field is the go func here.  So, IIUC, 'hour' could be a local variable within this function.
___
*[hopannotation/hopannotation.go, line 116 at r2](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-Mi1cS0F9nkeLjejDQ-W:-Mi2tXkP6L0L8bmm71j1:bsle3dv) ([raw file](https://github.com/m-lab/traceroute-caller/blob/b15eddf427b3a2cb9379e05fb143c34fb2e5f524/hopannotation/hopannotation.go#L116)):*
<details><summary><i>Previously, SaiedKazemi (Saied Kazemi) wrote…</i></summary><blockquote>

No, it's not backwards.  Hours go from 0 to 23 and then back to 0 again. Therefore, when the current hour is less than the previous hour, we have passed midnight.
</blockquote></details>

Ah got it.  Comments would be helpful.  I was misinterpreting.
___
*[hopannotation/hopannotation.go, line 111 at r3](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-Mi2spABCFVo0NDoxY8q:-Mi2spABCFVo0NDoxY8r:b-pkkbb2) ([raw file](https://github.com/m-lab/traceroute-caller/blob/eff8ec9dfc06b7a12359fa06bb71bd7124aff25f/hopannotation/hopannotation.go#L111)):*
> ```Go
> 	// at midnight.  Read tickerDuration and hour atomically to avoid
> 	// a race condition with package testing code.
> 	go func(tickerDuration int64) {
> ```

Can this parameter be a time.Duration?
___
*[hopannotation/hopannotation_test.go, line 1 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-Mi2uzID6QPG1A1KxKxD:-Mi2uzIE2jWbfMmU5scZ:b-29a1o8) ([raw file](https://github.com/m-lab/traceroute-caller/blob/ddba1bdfb3720a796d2f243a001ac440ba6fdf5a/hopannotation/hopannotation_test.go#L1)):*
> ```Go
> package hopannotation
> ```

I continue to be annoyed that Effective Go makes no suggestions about black box testing.  8-(
___
*[hopannotation/hopannotation_test.go, line 31 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-Mi3-73m65bIfTwdD_XM:-Mi3-73m65bIfTwdD_XN:b2x2lme) ([raw file](https://github.com/m-lab/traceroute-caller/blob/ddba1bdfb3720a796d2f243a001ac440ba6fdf5a/hopannotation/hopannotation_test.go#L31)):*
> ```Go
> 		{[]string{"1.2.3.4", "5.6.7.8"}, nil, 2, 2},                // should annotate and archive both
> 		{[]string{"1.2.3.4", "5.6.7.8"}, nil, 2, 2},                // should not annotate and archive either
> 		{[]string{"clear-cache", ""}, nil, 0, 0},                   // not a test, just clear the cache
> ```


But this is making my head explode a little, because I've always thought of the tests in a table as independent, rather than as a sequence of things that can alter state.  I think I like it though.

However, the use of the hop as a test directive seems a bit of a stretch.  Instead, I'd suggest adding a boolean 'clear_cache' indicating whether the hopcache should be cleared before proceeding.  

BTW: https://dave.cheney.net/2019/05/07/prefer-table-driven-tests suggests using a map from name to test struct, which I quite like.  (I'm still trying to adjust to table-driven being the preferred structure).
___
*[hopannotation/hopannotation_test.go, line 108 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-Mi2uXqD3tfgta9JSm9n:-Mi2uXqD3tfgta9JSm9o:brxnyks) ([raw file](https://github.com/m-lab/traceroute-caller/blob/ddba1bdfb3720a796d2f243a001ac440ba6fdf5a/hopannotation/hopannotation_test.go#L108)):*
> ```Go
> 	cancel()
> 	fakeMidnight(hc)
> 	if atomic.LoadInt32(&hc.hour) != 24+1 {
> ```

Ugh.  Now I see the race problem.

This would be much better written with monkey patch.  Then you can control the progression of time explicitly, or implicitly.  There are some subtleties, but it would avoid having to leak 'hour' outside the module, and would make the test code much simpler.

While testability often makes code more complicated, I'd really like to avoid additional complexity just to support testing, whenever possible.  I'm not crazy about monkey patch, but it has the advantage of keeping the complexity within the test code, and avoiding mucking up the production code.

Again, I'd be ok adding a TODO and issue, and fixing this later, particularly as it is intertwined with the hopcache race condition we talked about on the VC call earlier.


<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
___
*[hopannotation/hopannotation.go, line 116 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-Mhytw-wDoMibYFmAhGa:-Mi39IpDE9gLhe-5xApk:b2qiif8) ([raw file](https://github.com/m-lab/traceroute-caller/blob/ddba1bdfb3720a796d2f243a001ac440ba6fdf5a/hopannotation/hopannotation.go#L116)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Ok - had to think about what you meant for a bit.  Now I'm curious why there is a race in testing.  It seems like the only goroutine that should access this field is the go func here.  So, IIUC, 'hour' could be a local variable within this function.
</blockquote></details>

You've mentioned in a future comment that it's clear how hour is used to fake passage of midnight.


<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
___
*[hopannotation/hopannotation_test.go, line 31 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-Mi3-73m65bIfTwdD_XM:-Mi3C6nw9ZNJSdIMOfNK:b-896fix) ([raw file](https://github.com/m-lab/traceroute-caller/blob/ddba1bdfb3720a796d2f243a001ac440ba6fdf5a/hopannotation/hopannotation_test.go#L31)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>


But this is making my head explode a little, because I've always thought of the tests in a table as independent, rather than as a sequence of things that can alter state.  I think I like it though.

However, the use of the hop as a test directive seems a bit of a stretch.  Instead, I'd suggest adding a boolean 'clear_cache' indicating whether the hopcache should be cleared before proceeding.  

BTW: https://dave.cheney.net/2019/05/07/prefer-table-driven-tests suggests using a map from name to test struct, which I quite like.  (I'm still trying to adjust to table-driven being the preferred structure).

</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

gfr10598 reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/m-lab/traceroute-caller/120)* status: 1 change requests, 0 of 1 approvals obtained (waiting on @gfr10598 and @SaiedKazemi)

___
*[hopannotation/hopannotation.go, line 156 at r2](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-Mi1efuR4XxwzP6q9bzX:-Mi3CN6c2QIDJ6RISeMk:b-u8wodh) ([raw file](https://github.com/m-lab/traceroute-caller/blob/b15eddf427b3a2cb9379e05fb143c34fb2e5f524/hopannotation/hopannotation.go#L156)):*
<details><summary><i>Previously, SaiedKazemi (Saied Kazemi) wrote…</i></summary><blockquote>

We don't have to lock the map for the whole duration.  Please see my comment below.
</blockquote></details>

I don't mean that we have to, just that it would likely be more efficient.
___
*[hopannotation/hopannotation.go, line 170 at r4](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-Mi3Ah6V4rMcF6La-_eW:-Mi3Ah6V4rMcF6La-_eX:b-fy5djg) ([raw file](https://github.com/m-lab/traceroute-caller/blob/07ba82bebfcdf6608c60a4ad22c4ae2601143f67/hopannotation/hopannotation.go#L170)):*
> ```Go
> 	for _, hop := range hops {
> 		hc.hopsLock.Lock()
> 		if !hc.hops[hop+yyyymmdd] {
> ```

So, IIUC, this means that should yesterday's hops sneak into today's cache, they won't matter - they will be ignored when new traces with today's timestamp are checked.  Seems like this should address the issue.
TNHIWHDI
Thanks!


<!-- Sent from Reviewable.io -->

gfr10598 reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/m-lab/traceroute-caller/120)* status: 1 change requests, 0 of 1 approvals obtained (waiting on @gfr10598 and @SaiedKazemi)

___
*[hopannotation/hopannotation.go, line 116 at r2](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-Mi1cS0F9nkeLjejDQ-W:-Mi3H1HN7OytUluVMcpy:b-m77a30) ([raw file](https://github.com/m-lab/traceroute-caller/blob/b15eddf427b3a2cb9379e05fb143c34fb2e5f524/hopannotation/hopannotation.go#L116)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Ah got it.  Comments would be helpful.  I was misinterpreting.
</blockquote></details>

Just put this in a comment in the code or at the field definition.


<!-- Sent from Reviewable.io -->

gfr10598 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/m-lab/traceroute-caller/120)* status: 1 change requests, 0 of 1 approvals obtained (waiting on @SaiedKazemi)

___
*[hopannotation/hopannotation.go, line 156 at r2](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-Mi1efuR4XxwzP6q9bzX:-Mi3HFhu-JeVjb8vxw7N:bdb3bcg) ([raw file](https://github.com/m-lab/traceroute-caller/blob/b15eddf427b3a2cb9379e05fb143c34fb2e5f524/hopannotation/hopannotation.go#L156)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

I don't mean that we have to, just that it would likely be more efficient.
</blockquote></details>

To be clear, this is a nice to have - not required.
___
*[hopannotation/hopannotation.go, line 156 at r2](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-Mi1fL-0B47qGfZty_zq:-Mi3HNF4-84QosAkomXG:b-ux72ar) ([raw file](https://github.com/m-lab/traceroute-caller/blob/b15eddf427b3a2cb9379e05fb143c34fb2e5f524/hopannotation/hopannotation.go#L156)):*
<details><summary><i>Previously, SaiedKazemi (Saied Kazemi) wrote…</i></summary><blockquote>

If the cache is reset while we are iterating this loop, it means that midnight has passed.  Therefore, the remaining hops in the hops slice will be annotated which is exactly the behavior we want.

I will add a comment explaining this in the code.
</blockquote></details>

We discussed this in further detail on the VC.


<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
___
*[hopannotation/hopannotation.go, line 116 at r2](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-Mi1cS0F9nkeLjejDQ-W:-Mi3IPjq7HRnqst3YgvX:b-896fix) ([raw file](https://github.com/m-lab/traceroute-caller/blob/b15eddf427b3a2cb9379e05fb143c34fb2e5f524/hopannotation/hopannotation.go#L116)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Just put this in a comment in the code or at the field definition.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
___
*[hopannotation/hopannotation.go, line 156 at r2](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-Mi1efuR4XxwzP6q9bzX:-Mi3Idbs7qiXpGnFcBuu:b-896fix) ([raw file](https://github.com/m-lab/traceroute-caller/blob/b15eddf427b3a2cb9379e05fb143c34fb2e5f524/hopannotation/hopannotation.go#L156)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

To be clear, this is a nice to have - not required.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
___
*[hopannotation/hopannotation.go, line 156 at r2](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-Mi1fL-0B47qGfZty_zq:-Mi3IluGDZjp1VtbN_PG:by05yok) ([raw file](https://github.com/m-lab/traceroute-caller/blob/b15eddf427b3a2cb9379e05fb143c34fb2e5f524/hopannotation/hopannotation.go#L156)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

We discussed this in further detail on the VC.
</blockquote></details>

Made the hop cache time-aware, so we should never miss a hop annotation for today even if the traceroute ran at 23:59:59 and finished after midnight.


<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
___
*[hopannotation/hopannotation.go, line 170 at r4](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-Mi3Ah6V4rMcF6La-_eW:-Mi3JHBCA-NxyasVwsTe:b-lbejru) ([raw file](https://github.com/m-lab/traceroute-caller/blob/07ba82bebfcdf6608c60a4ad22c4ae2601143f67/hopannotation/hopannotation.go#L170)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

So, IIUC, this means that should yesterday's hops sneak into today's cache, they won't matter - they will be ignored when new traces with today's timestamp are checked.  Seems like this should address the issue.
TNHIWHDI
Thanks!
</blockquote></details>

What does TNHIWHDI mean?!


<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
___
*[hopannotation/hopannotation_test.go, line 108 at r1](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-Mi2uXqD3tfgta9JSm9n:-Mi3CFYr981RitGfSprd:bgnmhim) ([raw file](https://github.com/m-lab/traceroute-caller/blob/ddba1bdfb3720a796d2f243a001ac440ba6fdf5a/hopannotation/hopannotation_test.go#L108)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Ugh.  Now I see the race problem.

This would be much better written with monkey patch.  Then you can control the progression of time explicitly, or implicitly.  There are some subtleties, but it would avoid having to leak 'hour' outside the module, and would make the test code much simpler.

While testability often makes code more complicated, I'd really like to avoid additional complexity just to support testing, whenever possible.  I'm not crazy about monkey patch, but it has the advantage of keeping the complexity within the test code, and avoiding mucking up the production code.

Again, I'd be ok adding a TODO and issue, and fixing this later, particularly as it is intertwined with the hopcache race condition we talked about on the VC call earlier.
</blockquote></details>

Ack'ed.  I have added a TODO to use money patch in a later PR.


<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
___
*[hopannotation/hopannotation.go, line 111 at r3](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-Mi2spABCFVo0NDoxY8q:-Mi38ovI3kqExZ92fIHH:b-896fix) ([raw file](https://github.com/m-lab/traceroute-caller/blob/eff8ec9dfc06b7a12359fa06bb71bd7124aff25f/hopannotation/hopannotation.go#L111)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Can this parameter be a time.Duration?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

Commit:
Make additional changes suggested in code review

gfr10598 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-:-Mi3fNpi5TTE-8ywTPpk:bnfp4nl)* status: :shipit: complete! 1 of 1 approvals obtained

___
*[hopannotation/hopannotation.go, line 170 at r4](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-Mi3Ah6V4rMcF6La-_eW:-Mi3f1yI4WWi2y0yLY6K:b-tlos7l) ([raw file](https://github.com/m-lab/traceroute-caller/blob/07ba82bebfcdf6608c60a4ad22c4ae2601143f67/hopannotation/hopannotation.go#L170)):*
<details><summary><i>Previously, SaiedKazemi (Saied Kazemi) wrote…</i></summary><blockquote>

What does TNHIWHDI mean?!
</blockquote></details>

That's not how I would have done it.  I think it is from Extreme Programming.  I didn't mean to leave that in - got distracted and sent by mistake.


<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
Thanks for your time and review.





<!-- Sent from Reviewable.io -->

SaiedKazemi commented:
___
*[hopannotation/hopannotation.go, line 170 at r4](https://reviewable.io/reviews/m-lab/traceroute-caller/120#-Mi3Ah6V4rMcF6La-_eW:-Mi3hPpQEY53fhaeNBpx:bl7thbq) ([raw file](https://github.com/m-lab/traceroute-caller/blob/07ba82bebfcdf6608c60a4ad22c4ae2601143f67/hopannotation/hopannotation.go#L170)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

That's not how I would have done it.  I think it is from Extreme Programming.  I didn't mean to leave that in - got distracted and sent by mistake.
</blockquote></details>

LOL :). I googled it, but nothing came up!


<!-- Sent from Reviewable.io -->
